pulser/pulse.py

```diff
@@ -69,27 +70,35 @@ def __new__(cls, *args, **kwargs):  # type: ignore
         else:
             return object.__new__(cls)
 
-    def __post_init__(self) -> None:
+    def __init__(
+        self,
+        amplitude: Union[Waveform, Parametrized],
+        detuning: Union[Waveform, Parametrized],
+        phase: Union[float, Parametrized],
+        post_phase_shift: Union[float, Parametrized] = 0.0,
+    ):
         """Initializes a new Pulse."""
         if not (
-            isinstance(self.amplitude, Waveform)
-            and isinstance(self.detuning, Waveform)
+            isinstance(amplitude, Waveform) and isinstance(detuning, Waveform)
         ):
             raise TypeError("'amplitude' and 'detuning' have to be waveforms.")
 
-        if self.detuning.duration != self.amplitude.duration:
+        if detuning.duration != amplitude.duration:
             raise ValueError(
                 "The duration of detuning and amplitude waveforms must match."
             )
-        if np.any(self.amplitude.samples < 0):
+        if np.any(amplitude.samples < 0):
             raise ValueError(
                 "All samples of an amplitude waveform must be "
                 "greater than or equal to zero."
             )
-
-        object.__setattr__(self, "phase", self.phase % (2 * np.pi))
+        object.__setattr__(self, "amplitude", amplitude)
+        object.__setattr__(self, "detuning", detuning)
+        phase = cast(float, phase)
+        object.__setattr__(self, "phase", float(phase) % (2 * np.pi))
+        post_phase_shift = cast(float, post_phase_shift)
         object.__setattr__(
-            self, "post_phase_shift", self.post_phase_shift % (2 * np.pi)
+            self, "post_phase_shift", float(post_phase_shift) % (2 * np.pi)
```

Comment:
Yeah, I know it's pretty ugly... It's because the `dataclass` is frozen, which means that calls to its `__setattr__` are blocked, so you have to do this in this hacky way.

Reply:
Ah okay thanks!